### DISCUSSION TITLE
Why Rust as a backend for PydanticV2?

### DISCUSSION THREAD

**Discussion by vdksoda** [pydantic/pydantic#5279] [Pydantic V2]
Thanks to the Pydantic team for the details on the PydanticV2 roadmap and being forward in discussing PydanticV2 with the community early on. The vision is exciting!

I have no knowledge on Rust but I am aware of Cython. Pydantic V1 has some portion of code in Cython as well. I was hoping someone could share some resource/points on why Rust was preferred to Cython -specifically for Pydantic's functionality. 

This is merely for knowledge and information. In my case I would use the response as a starting point to become more knowledgable on pros and cons of Rust for a usecase (Pydantic) I use regularly. This is very different to generic pros and cons of Rust that one could find on the internet.

**Comment by samuelcolvin** [comment]
First of all, cython can mean two things:
* a language (and compiler) for relatively performant python extensions - it's (in my limited experience) somewhat akin to writing c code with pythoneque syntax, sample [here](https://cython.readthedocs.io/en/latest/src/quickstart/cythonize.html#id3) - note: this is **not** what pydantic v1 used
* using that same compiler to slightly (~50%) speedup a subset of vanilla python code - this is what pydantic V1 did - used the cython compiler to compile code which could also run with cpython

---

To answer the question, there are three reason I chose to write pydantic V2 in Rust:
* **performance** rust is massively faster than cpython, python extensions built with rust (via pyo3) are therefore significantly faster than python extensions written in python and generally faster than statically typed cython extensions
* **safety**, compared to python: rust forces you to account for all possible extensions, compared to c: rust makes it much harder to introduce memory errors
* **maintainability** pydantic-core written in rust should be significantly easier to maintain for the next 10 years than the equivilant code written in python, it should also be much easier to get started with than statically typed cython which has a much smaller user base

**Comment by vdksoda** [comment]
Thank you for your speedy response.
